### ISSUE TITLE
Middleware `response` is not invoked if error result was returned by another middleware's `start` 

### ISSUE THREAD

**Issue by mitskevich** [actix/actix-web#255]
actix-web 0.6.10

Example:

```rust
extern crate actix_web;

use actix_web as aw;
use actix_web::middleware::{Middleware, Response, Started};

struct MiddlewareOne;

impl<S> Middleware<S> for MiddlewareOne {
    fn start(&self, _req: &mut aw::HttpRequest<S>) -> Result<Started, aw::Error> {
        println!("middleware one - start");
        Err(aw::error::ErrorBadRequest("error msg"))
    }
}

struct MiddlewareTwo;

impl<S> aw::middleware::Middleware<S> for MiddlewareTwo {
    fn response(
        &self,
        _req: &mut aw::HttpRequest<S>,
        resp: aw::HttpResponse,
    ) -> Result<Response, aw::Error> {
        println!("middleware two - response");
        Ok(Response::Done(resp))
    }
}

fn index(_r: aw::Path<()>) -> String {
    format!("hello")
}

fn main() {
    aw::server::new(|| {
        aw::App::new()
            .middleware(MiddlewareOne)
            .middleware(MiddlewareTwo)
            .resource("/index/", |r| r.with(index))
    }).bind("127.0.0.1:8080")
        .unwrap()
        .run();
}
```

Here I'm registering two middlewares. First one is throwing an error in its `start` method. Second one has `response` method that just prints to the console. 

I expect that `MiddlewareTwo::response` should always be invoked. When I call `curl -i http://localhost:8080/index/` I get correct `400` status code, but console output contains only one line:

```
middleware one - start
```

I assume that `response` was not invoked at all.

Because of that behaviour, middlewares like `ErrorHandlers` or `Logger` are not working correctly when any middleware's `start` method returns an error.

**Comment by fafhrd91** [comment]
this is expected behavior. middleware error halts request handling.

**Comment by mitskevich** [comment]
OK, a few questions then:

1. `Logging` middleware prints to `stdout` inside its `response()`. If some middleware returns `actix_web::error::Error*` error from its `start()` method, user receives actual `4xx` or `5xx` response from the server, but that request/response is never logged to `stdout`. Is that a correct behaviour? Let me know if my explanation is confusing, I can provide a concrete code snippet example.

2. Similarly, `ErrorHandlers` middleware doesn't catch errors that are returned by `start()` methods of other middlewares. Is that how it should work too? Again, I can write up some example code, let me know if you need it.

3. Initially I wanted to create a middleware that should process user input (common cookies and data payload for all my requests) and store it inside request extension. That middleware has to be able to return `4xx` statuses in case if user provides incorrect data. But if this behaviour will break other middlewares, I can't do that. Do I understand correctly, that middlewares shouldn't be used for cases like that, and instead I should use wrappers for all my handlers (or something else)?

**Comment by fafhrd91** [comment]
for 1 and 2, depends on middleware order. if middleware's start method get called then response method should be call too. 

3, your middleware should be reregistered last

**Comment by mitskevich** [comment]
1. Thanks for clarification! It seems, I was wrong about `Logging` it actually logs in `finish()` and it depends on ordering, so my first question is obsolete...

2. ...but `ErrorHandlers` is never invoked for error returned by `Middleware::start()` independent of middleware order:

```rust
extern crate actix_web;

use actix_web as aw;
use actix_web::middleware::{Middleware, Response, Started};

struct MiddlewareOne;

impl<S> Middleware<S> for MiddlewareOne {
    fn start(&self, _req: &mut aw::HttpRequest<S>) -> Result<Started, aw::Error> {
        Err(aw::error::ErrorInternalServerError("middleware error"))
    }
}

fn index(_r: aw::Path<()>) -> Result<String, aw::Error> {
    Err(aw::error::ErrorInternalServerError("request error"))
}

fn render_500<S>(_: &mut aw::HttpRequest<S>, resp: aw::HttpResponse) -> aw::Result<Response> {
    println!("render_500 handler invoked!");
    Ok(Response::Done(resp))
}

fn main() {
    aw::server::new(|| {
        aw::App::new()
            .middleware(aw::middleware::ErrorHandlers::new()
                .handler(aw::http::StatusCode::INTERNAL_SERVER_ERROR, render_500))
            .middleware(MiddlewareOne)
            .resource("/index/", |r| r.with(index))
    }).bind("127.0.0.1:8080")
        .unwrap()
        .run();
}
```

This server never invokes `render_500` function, even if I change ordering of middlewares.

3. If I change middleware ordering in my first example (the one in the opening of this issue), it never invokes `MiddlewareTwo::response()`, so my middleware will actually break other middlewares, even if I'll register it last.

Sorry if I'm just misunderstanding something and wasting your time with these questions.

**Comment by fafhrd91** [comment]
ok, I'll check 2 and 3

**Comment by joshleeb** [comment]
@mitskevich the reason for Part 3 is because the middleware functions are run in this order (with the following `main` function):

```rust
fn main() {
    aw::server::new(|| {
        aw::App::new()
            .middleware(MiddlewareTwo)
            .middleware(MiddlewareOne)
            .resource("/index/", |r| r.with(index))
    }).bind("127.0.0.1:8080")
        .unwrap()
        .run();
}
```

1. `MiddlewareTwo::start()`
2. `MiddlewareOne::start()`
	- `ErrorBadRequest` is returned.
3. `MiddlewareTwo::response()` (would be run next)
4. `MiddlewareOne::response()` (would be run next)

This would also explain Part 2. `ErrorHandlers` is an implementation of `Middleware` and only defines the `Middleware:::response()` function. So, as we are just bailing out of the pipeline whenever a middleware function returns `Err`, the error being returned is never passed to any of the `ErrorHandlers`.

IMHO it's more intuitive to run through all of `MiddlewareTwo`, that is `start()`, `response()`, and `finish()`, before moving onto `MiddlewareOne`. 

Then, if an error is encountered at any of these stages, we bail out of executing any of the other middleware and jump straight to the `ErrorHandlers` middleware

@fafhrd91 wdyt?

**Comment by fafhrd91** [comment]
Original idea was to run .response() method in backward order. Also we should call .finish() immediately because .response() is called when Httpresponse is ready
But we still need to process response body, which could be async operation.

**Comment by fafhrd91** [comment]
fixed in master and 0.6 

i'll release new 0.6

**Comment by mitskevich** [comment]
thanks!
